Parquet Incremental Sync

sapienza88 opened the pull request:
## What is the purpose of the pull request

Adds incremental syncing ability to the ParquetSource

## Brief change log

- Adds a new class [ParquetDataManager.java](https://github.com/apache/incubator-xtable/pull/768/changes#diff-bf0010f99dc76fa978ff394f9dd262cb8e75fd71848a56ab8ecc48f9621e9a99) for handling the fetching of data files for Parquet Source
- Updates IT to include incremental source

## Verify this pull request

- new tests added to ITParquetConversionSource

Commit:
given a parquet file return data from a certain modification time

Commit:
create the path based on the partition then inject the file to append into the parquet table

Commit:
Handle case of path construction with file partitioned over many fields, interfacing with ConversionSource

Commit:
test append Parquet file into table init

Commit:
add function to test schema equivalence before appending

Commit:
construct path to inject to based on partitions

Commit:
fix imports

vinishjail97 commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetDataManager.java:
```diff
@@ -0,0 +1,267 @@
+public class ParquetDataManager {
+  private ParquetMetadataExtractor metadataExtractor = ParquetMetadataExtractor.getInstance();
+
+  public ParquetReader readParquetDataAsReader(String filePath, Configuration conf) {
```
This is just returning the ParquetReader and not reading any data? Can we rename to getParquetReader?

vinishjail97 commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetDataManager.java:
```diff
@@ -0,0 +1,267 @@
+    try {
+      reader = ParquetReader.builder(new GroupReadSupport(), file).withConf(conf).build();
+    } catch (IOException e) {
+      e.printStackTrace();
```
Avoid e.printStackTrace() and use logger instead.

vinishjail97 commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetDataManager.java:
```diff
@@ -0,0 +1,267 @@
+      }
+    } catch (Exception e) {
+
+      e.printStackTrace();
```
Use logger?

vinishjail97 commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetDataManager.java:
```diff
@@ -0,0 +1,267 @@
+            conf)) {
+      Group currentGroup = null;
+      while ((currentGroup = (Group) reader.read()) != null) {
+        writer.write(currentGroup);
```
Why are we writing new parquet files again like this through the writer? I think there's some misunderstanding with the parquet incremental sync feature here. 

Parquet Incremental Sync Requirements.
1. You have a target table where parquet files [p1/f1.parquet, p1/f2.parquet, p2/f1.parquet] have been synced to hudi, iceberg and delta for example. 
2. In the source changes some changes have been made a new file in partition p1 was added and p2's file was deleted. The incremental sync should now sync the new changes incrementally. 

@sapienza88 It's better to align on the approach first here before we push PR's. Can you add the approach for parquet incremental sync in the PR description or any google doc if possible?

sapienza88 replied:
@vinishjail97 thanks for the suggestion, but that isn't helping. Could you elaborate on that idea and how you could manage the metadata only for the task of retrieving data from a (modification) particular date? at the very least the current ConversionSource wasn't coded with that in mind.

Commit:
refactoring (lombok, logs, javadocs and function and approach commenting)

Commit:
use appendFile to append a file into a table while tracking the appends using the FileStatus' modifTime attribute

Commit:
find the files that satisfy to the time condition

Commit:
treat appends as separate files to add in the target partition folder

Commit:
update approach: selective block compaction

Commit:
update approach: added a basic test to check data selection using modificationTime selector

Commit:
fix append based on partition value

Commit:
fix test with basic example where partitions are not considered

Commit:
fix test with basic example where partitions are not considered2

Commit:
fix test with basic example where partitions are not considered3

Commit:
test with time of last append is now

Commit:
test appendFile with Parquet: TODO test with multiple partitions 1) append and 2) filter for sync

Commit:
merge recursively one partition files

Commit:
fix paths for files to append

Commit:
fix bug of appending file path

Commit:
fix bug of schema

Commit:
make sure returned files are non empty when iterating in the partitions

Commit:
make sure returned files are non empty when iterating in the partitions 2

Commit:
discard empty files when appending

Commit:
discard empty files when appending

Commit:
one spark session for both tests

Commit:
bug fix

Commit:
bug fix

Commit:
cleanups + TODO: partitions match and merge

Commit:
added test for many partitions, TODO integrate functions into ParquetConversionSource

Commit:
selecting data bug fix, TODO integrate functions into ParquetConversionSource

Commit:
run all tests, TODO integrate functions into ParquetConversionSource

Commit:
run all tests, TODO integrate functions into ParquetConversionSource

Commit:
spotless:apply, TODO integrate functions into ParquetConversionSource

Commit:
bug fixes, TODO integrate functions into ParquetConversionSource

Commit:
bug fixes, TODO integrate functions into ParquetConversionSource

vinishjail97 commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetDataManager.java:
```diff
@@ -0,0 +1,325 @@
+    return parquetFileConfig;
+  }
+  // TODO add safe guards for possible empty parquet files
+  // append a file (merges two files into one .parquet under a partition folder)
```
Why are we merging two parquet files to one? We are not building a compaction service right?

vinishjail97 reviewed (commented):
@sapienza88 Should I push a scaffolding PR for basic functionality with incremental sync and you can take it forward from there?

sapienza88 commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetDataManager.java:
```diff
@@ -0,0 +1,325 @@
+            DEFAULT_BLOCK_SIZE,
+            0);
+    // write the initial table with the appended file to add into the outputPath
+    writer.start();
```
yes writing/copying data to append it

sapienza88 commented:
@vinishjail97 you mentioned this earlier "Filters files by modification time", how that time is set when the files are ingested?in the footers?

Commit:
handle lightweight metadata in the conversionSource

sapienza88 commented:
@vinishjail97 @the-other-tim-brown pls check the latest changes for a review. thanks.

Commit:
fix CI

Commit:
fix CI, refactoring and isIncrementalSyncSafeFrom implementation

Commit:
added test

Commit:
added test: init conversionSource instance

Commit:
added test: init conversionSource instance: fix error

Commit:
added test: fix error

Commit:
more tests + bug fixes and reformatting

Commit:
CI fix

Commit:
CI fix

Commit:
CI fix

Commit:
CI fix

Commit:
CI fix

Commit:
Merge branch 'main' of https://github.com/sapienza88/incubator-xtable into parquet_incr_sync

Commit:
Merge branch 'main' of https://github.com/sapienza88/incubator-xtable into parquet_incr_sync

Commit:
CI fix

Commit:
merge ready fixes & tests + refactoring & reformatting

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetFileConfig.java:
```diff
@@ -0,0 +1,70 @@
+
+  public ParquetFileConfig(Configuration conf, Path file) {
+    long modifTime = -1L;
+    ParquetMetadata metadata =
```
Reading the metadata for all files is going to be quite expensive. Let's lazily evaluate this when it is needed?

Commit:
optimization: read metadata once for target file, read parquet files once, reformatting

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/parquet/ITParquetDataManager.java:
```diff
@@ -0,0 +1,208 @@
+import org.apache.xtable.model.TableChange;
+import org.apache.xtable.model.storage.TableFormat;
+
+public class ITParquetDataManager {
```
This test is not directly testing the ParquetDataManager. It is testing the ParquetConversionSource and therefore the tests can be moved in with the ITParquetConversionSource

the-other-tim-brown committed:
Lazily fetch info, move state into ParquetDataManager

the-other-tim-brown committed:
add TestParquetDataManager, simplify data manager methods

the-other-tim-brown committed:
make IT parameterized, add incremental sync

the-other-tim-brown committed:
fix schema extraction to fix snapshot sync

the-other-tim-brown committed:
add optimizations

Commit:
fix for ParquetDataManager Test (mock())

Commit:
before resycing delete the first synced files metadata

Commit:
in order to append parquet files, perform union then spark overwrite (not append)

Commit:
solving directory issue when overwriting data using spark

Commit:
solving directory issue when overwriting data using spark2

Commit:
solving directory issue when overwriting data using spark3

Commit:
solving directory issue when overwriting data using spark4

Commit:
solving directory issue when overwriting data using spark5

Commit:
solving directory issue when overwriting data using spark6

Commit:
solving directory issue when overwriting data using spark7

Commit:
revert writeData changes

Commit:
solving directory issue when overwriting data using spark8

Commit:
solving directory issue when overwriting data using spark9

Commit:
solving directory issue when overwriting data using spark9

Commit:
revert changes

Commit:
testing without cleaning metadata for the second sync

Commit:
testing without cleaning metadata for the second sync

Commit:
testing without cleaning metadata for the second sync

Commit:
testing without cleaning metadata for the second sync

Commit:
testing without cleaning metadata for the second sync

Commit:
revert changes

Commit:
revert changes

Commit:
revert changes

Commit:
use no tempDir

Commit:
revert changes

Commit:
use union() to append data for the second sync()

Commit:
use union() to append data for the second sync()

Commit:
use union() to append data for the second sync()

Commit:
use union() to append data for the second sync()

Commit:
use union() to append data for the second sync() reverted

sapienza88 commented on xtable-core/src/test/java/org/apache/xtable/parquet/ITParquetConversionSource.java:
```diff
@@ -184,77 +200,43 @@ private ConversionSourceProvider<?> getConversionSourceProvider(String sourceTab
-          new ConversionController(jsc.hadoopConfiguration());
-      conversionController.sync(conversionConfig, conversionSourceProvider);
-      checkDatasetEquivalenceWithFilter(sourceTableFormat, tableToClose, targetTableFormats, false);
+  private void cleanupTargetMetadata(String dataPath, List<String> formats) {
```
for now I don't see why the second sync is not considering the first one as done, it takes the first 5 rows plus the new data write (5 rows + 1 appended) as the data to sync (while it should take only (5 rows + 1 appended))

cleanupTargetMetadata() is only to bypass the error...

the-other-tim-brown replied:
The test should not be updated to hide the bug. The bug needs to be fixed in the feature.

Commit:
fix second sync error

Commit:
fix second sync error

Commit:
fix second sync error

Commit:
fix second sync error

Commit:
fix second sync error

Commit:
fix second sync error

Commit:
fix second sync error

Commit:
fix second sync error

Commit:
fix second sync error

Commit:
fix second sync error

Commit:
fix second sync error

Commit:
remove old blocking tests

Commit:
solving conflicts

Commit:
spotless:apply and more tests

Commit:
spotless:apply and test fix

Commit:
spotless:apply and test fix

Commit:
spotless:apply and test fix

Commit:
test fix

Commit:
spotless

Commit:
spotless imports

Commit:
spotless imports

Commit:
add syncMode Incr

Commit:
revert changes

Commit:
add syncMode Incr

Commit:
spotless:apply

the-other-tim-brown committed:
update naming

the-other-tim-brown committed:
minimize diff with main

vinishjail97 commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetConversionSource.java:
```diff
@@ -162,14 +167,17 @@ public TableChange getTableChangeForCommit(Long modificationTime) {
             .filter(fileStatus -> fileStatus.getModificationTime() > modificationTime)
             .collect(Collectors.toList());
-    InternalTable internalTable = getMostRecentTable(parquetFiles);
+    InternalTable internalTable = getMostRecentTable(getParquetFiles(hadoopConf, basePath));
```
Boundary inconsistency nearby (line 168 filter): this method uses `fileStatus.getModificationTime() > modificationTime`, but `ParquetDataManager.getParquetFilesMetadataAfterTime(long)` filters with `>=`. A file whose mod time equals the prior sync instant is silently dropped from the diff here. Pick one semantic (ideally `>` everywhere — the instant is exclusive) and apply consistently across both classes?

sapienza88 replied:
ok this is adjusted now

vinishjail97 commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetDataManager.java:
```diff
@@ -0,0 +1,108 @@
+    this.hadoopConf = hadoopConf;
+    this.basePath = basePath;
+    try {
+      URI uri = Paths.get(basePath).toUri();
```
Blocking — `Paths.get(basePath).toUri()` uses the JDK's *default* FileSystemProvider (local filesystem). For `s3://bucket/...`, `gs://...`, `abfs://...`, this either throws `InvalidPathException` (colon is invalid on some platforms) or silently produces a `file://` URI, after which `FileSystem.get(uri, hadoopConf)` returns the local Hadoop FS — not S3A/GCS/ABFS. That breaks all cloud storage users.

Use Hadoop's `Path` instead: `URI uri = new Path(basePath).toUri();`. This mirrors how Hudi and the rest of xtable resolve base paths for cloud stores.

sapienza88 replied:
I used that instead now

vinishjail97 commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetDataManager.java:
```diff
@@ -0,0 +1,108 @@
+ * parquet files and filtering the files based on the modification times.
+ */
+@Log4j2
+@RequiredArgsConstructor
```
Nit — exposing both the Lombok-generated 3-arg ctor (`@RequiredArgsConstructor`) and an explicit 2-arg ctor creates ambiguity about the public API. Production code calls the 2-arg form; tests call the 3-arg form. Consider dropping `@RequiredArgsConstructor` and either annotating the 3-arg ctor `@VisibleForTesting` or using a package-private static factory method for tests.

sapienza88 replied:
I'll leave the rest of your comments here to somebody else to do them.

the-other-tim-brown replied:
I fixed this to help unblock this

vinishjail97 commented on xtable-core/src/test/java/org/apache/xtable/parquet/ITParquetConversionSource.java:
```diff
@@ -418,6 +432,124 @@ private void checkDatasetEquivalence(
+            RowFactory.create(103, "BA", 2027, 11));
+
+    Dataset<Row> dfInit = sparkSession.createDataFrame(data, schema);
+    Path fixedPath = Paths.get("target", "fixed-parquet-data", "parquet_table_test_2");
```
Relative path `Paths.get("target", "fixed-parquet-data", "parquet_table_test_2")` pollutes the workspace across test runs, isn't cleaned up, and makes the test order-dependent when re-run without `./gradlew clean`. This class already uses the `@TempDir` pattern — please use it here too. Also drop the commented-out `// String outputPath = fixedPath.toString();` on line 457.

the-other-tim-brown replied:
Updated this

vinishjail97 commented on xtable-core/src/test/java/org/apache/xtable/parquet/ITParquetConversionSource.java:
```diff
@@ -418,6 +432,124 @@ private void checkDatasetEquivalence(
+    Instant instantBeforeFirstSnapshot =
+        Instant.ofEpochMilli(snapshot.getTable().getLatestCommitTime().toEpochMilli());
+    assertEquals(instantBeforeFirstSnapshot.toEpochMilli(), newModificationTime);
+    assertTrue(conversionSource.isIncrementalSyncSafeFrom(Instant.ofEpochMilli(testTime)));
```
This test asserts `changes` is non-null and `isIncrementalSyncSafeFrom(testTime) == true`, but does not validate the actual `filesAdded` set? 

Please build the expected set of physical paths (or `InternalDataFile`s) and `assertEquals` against `changes.getFilesDiff().getFilesAdded()`. Also add a negative case: `isIncrementalSyncSafeFrom` returns `false` when earliest file is newer than the instant, and when the table is empty. Without these, the test would pass even if the diff were silently empty.

sapienza88 replied:
added a test here

vinishjail97 commented on xtable-core/src/test/java/org/apache/xtable/parquet/TestParquetDataManager.java:
```diff
@@ -0,0 +1,453 @@
+    assertNotNull(result);
+    List<ParquetFileInfo> fileList = result.collect(Collectors.toList());
+    assertEquals(3, fileList.size());
+    assertEquals(1000L, fileList.get(0).getModificationTime());
```
Nit — asserting positional ordering here relies on `RemoteIterator` + `Collectors.toList()` preserving the mock insertion order. In production, FS listing order is platform-dependent and not guaranteed. Either sort inside `getCurrentFilesInfo()` (and document the ordering contract) or switch these assertions to `Set<Long>` comparisons. Same pattern applies to `testGetParquetFilesMetadataAfterTime_someMatch` and `_exactTimeMatch`.

the-other-tim-brown replied:
I'm updating the test since sorting in the API requires you to pull everything into memory before sorting.

Commit:
added test for tableChangeAddedFiles and changed path construction of ParquetDataManager and adjusted operator for mostRecentFile + spotless

Commit:
added test for tableChangeAddedFiles and changed path construction of ParquetDataManager and adjusted operator for mostRecentFile + spotless + import error fixed

Commit:
tests fixed

Commit:
tests fixed

Commit:
tests fixed

Commit:
log for actual changes

Commit:
log for actual changes

Commit:
tests fixed

Commit:
spotless

Commit:
removed comment

Commit:
spotless

the-other-tim-brown committed:
address PR comments

the-other-tim-brown committed:
fix path

the-other-tim-brown committed:
add safety check around handling of out of sync targets with parquet source

vinishjail97 commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetConversionSource.java:
```diff
@@ -162,14 +182,17 @@ public TableChange getTableChangeForCommit(Long modificationTime) {
             .filter(fileStatus -> fileStatus.getModificationTime() > modificationTime)
             .collect(Collectors.toList());
-    InternalTable internalTable = getMostRecentTable(parquetFiles);
+    InternalTable internalTable = getMostRecentTable(getParquetFiles(hadoopConf, basePath));
```
`getTableChangeForCommit` lists the filesystem twice here (L178 and L185), and `getCurrentCommitState`/`getCurrentSnapshot` repeat the pattern (L208/220/222/227). Each `getParquetFiles(hadoopConf, basePath)` is a full recursive `listFiles` — on S3/GCS that is many round-trips. This is exactly @the-other-tim-brown's earlier "fetch the list once" comment, and it's still open. You already inject `parquetDataManager`, which caches the listing lazily (`@Getter(lazy=true)`) — please route all listings through it and drop the static `getParquetFiles` duplicate so there's a single source of truth.

vinishjail97 replied:
Addressed in 2170d361 — `getTableChangeForCommit` and `getCurrentSnapshot` now materialize the listing once and stream from it. Verified: TestParquetDataManager 10/10, ITParquetConversionSource 6/6. Resolving.

vinishjail97 commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetConversionSource.java:
```diff
@@ -162,14 +182,17 @@ public TableChange getTableChangeForCommit(Long modificationTime) {
     return TableChange.builder()
+        .sourceIdentifier(
+            getCommitIdentifier(
+                parquetDataManager.getMostRecentParquetFile().getModificationTime()))
```
The `sourceIdentifier` is derived from `parquetDataManager.getMostRecentParquetFile()`, but the table snapshot and `tableChangesAfter` come from a *separate* `getParquetFiles(...)` listing above. If a file lands between the two listings, the committed source identifier can advance past a file that wasn't included in `filesAdded`, marking that file permanently synced. Take one snapshot of the file list and derive both the changes and the identifier from it.

vinishjail97 replied:
Confirmed the gap was still present after the fetch-once cleanup — `tableChangesAfter`/`internalTable` used the materialized snapshot, but `sourceIdentifier` still called `parquetDataManager.getMostRecentParquetFile()`, a separate listing. Fixed in d510e1a9: both the table and the identifier now come from the same `parquetFiles` snapshot (`getMostRecentParquetFile(parquetFiles.stream())`). Added an assertion in ITParquetConversionSource that the committed identifier equals that snapshot's latest modification time. Verified: ITParquetConversionSource 6/6. Resolving.

vinishjail97 commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetDataManager.java:
```diff
@@ -0,0 +1,111 @@
+    return new ParquetFileInfo(hadoopConf, file);
+  }
+
+  ParquetFileInfo getParquetDataFileAt(long targetTime) {
```
`getParquetDataFileAt(long)` (and `getParquetFilesMetadataAfterTime(long)` at L105) are only referenced from `TestParquetDataManager` — no production caller. Either wire them into the incremental path (they look like the intended replacement for the duplicate listing in `ParquetConversionSource`) or remove them so we don't ship test-only API.

vinishjail97 replied:
Addressed in 2170d361 — removed `getParquetDataFileAt` and `getParquetFilesMetadataAfterTime` (no production caller) plus their tests; repointed the real-FS tests to `getCurrentFilesInfo`. Resolving.

vinishjail97 committed:
Address review feedback: fetch parquet listing once; drop unused ParquetDataManager methods

- ParquetConversionSource: materialize the file listing once per operation in
  getTableChangeForCommit and getCurrentSnapshot instead of re-listing the
  filesystem 2-3 times (addresses the 'fetch the list once' review comment).
- ParquetDataManager: remove getParquetDataFileAt and
  getParquetFilesMetadataAfterTime; both had no production caller (only tests).
- TestParquetDataManager: drop tests for the removed methods and repoint the
  real-filesystem tests to getCurrentFilesInfo.

vinishjail97 commented on xtable-core/src/main/java/org/apache/xtable/parquet/ParquetConversionSource.java:
```diff
@@ -149,87 +148,114 @@ public CommitsBacklog<Long> getCommitsBacklog(InstantsForIncrementalSync syncIns
+    List<ParquetFileConfig> filesMetadata =
+        parquetDataManagerExtractor.getParquetFilesMetadataAfterTime(
+            hadoopConf,
+            parquetDataManagerExtractor.getParquetFiles(hadoopConf, basePath),
```
Addressed in 2170d361. `getTableChangeForCommit` and `getCurrentSnapshot` now materialize the parquet listing once per operation (`getParquetFiles(...).collect(toList())`) and stream from that list, instead of re-listing the filesystem 2–3 times. Verified with `TestParquetDataManager` (10/10) and `ITParquetConversionSource` (6/6).

vinishjail97 committed:
Derive source identifier from the same listing as the table change

getTableChangeForCommit computed filesAdded and the table from one
materialized listing but derived the committed sourceIdentifier from a
separate parquetDataManager.getMostRecentParquetFile() listing. A file
landing between the two listings could advance the committed identifier
past a file that was never included in filesAdded, permanently skipping
it on the next incremental sync.

Use the most recent file from the same snapshot for both the table and
the source identifier. Add an assertion in ITParquetConversionSource
that the committed identifier matches that snapshot's latest mod time.
